lib/SILOptimizer/Mandatory/PhiStorageOptimizer.cpp

```diff
@@ -262,39 +293,100 @@ void PhiStorageOptimizer::tryCoalesceOperand(SILBasicBlock *incomingPred) {
   coalescedPhi.coalescedOperands.push_back(incomingOper);
 }
 
-// Record liveness generated by uses of \p incomingVal.
+/// The block which dominates all the defs whose storage is projected out of the
+/// storage for \p incomingVal.
+///
+/// Populates \p projectedValues with the values whose storage is recursively
+/// projected out of the storage for incomingVal.
+SILBasicBlock *
+PhiStorageOptimizer::getDominatingBlock(SILValue incomingVal,
+                                        StackList<SILValue> *projectedValues) {
+  assert(domInfo);
+
+  // Recursively find the set of value definitions and the dominating LCA.
+  ValueWorklist values(incomingVal->getFunction());
+
+  values.push(incomingVal);
+
+  SILBasicBlock *lca = incomingVal->getParentBlock();
+  auto updateLCA = [&](SILBasicBlock *other) {
+    lca = domInfo->findNearestCommonDominator(lca, other);
+  };
+
+  while (auto value = values.pop()) {
+    if (projectedValues)
+      projectedValues->push_back(value);
+    auto *defInst = value->getDefiningInstruction();
+    if (!defInst) {
+      assert(!PhiValue(value));
```

Comment:
Mysterious assert alert. It looks like it's saying that `incomingVal` can never be a phi. Where do we guarantee that?

Reply:
Yeah, this is a result of
```
  if (PhiValue(incomingVal)) {
    // Do not coalesce a phi with other phis. This would require liveness
    // analysis of the whole phi web before coalescing phi operands.
    return nullptr;
  }
```
We should be able to remove this bail-out (and add all values incoming to `PhiValue(value)` to the `values` `ValueWorklist` here) in a follow-up.